pkg/apis/kops/util/versions_test.go

```diff
@@ -21,21 +21,54 @@ import (
 )
 
 func Test_ParseKubernetesVersion(t *testing.T) {
-	grid := map[string]string{
-		"1.3.7":         "1.3.7",
-		"v1.4.0-beta.8": "1.4.0-beta.8",
-		"1.5.0":         "1.5.0",
-		"https://storage.googleapis.com/kubernetes-release-dev/ci/v1.4.0-alpha.2.677+ea69570f61af8e/": "1.4.0",
+	cases := []struct {
+		version       string
+		expected      string
+		expectedError bool
+	}{
+		{
+			version:       "1.3.7",
+			expected:      "1.3.7",
+			expectedError: false,
+		},
+		{
+			version:       "v1.4.0-beta.8",
+			expected:      "1.4.0-beta.8",
+			expectedError: false,
+		},
+		{
+			version:       "1.5.0",
+			expected:      "1.5.0",
+			expectedError: false,
+		},
+		{
+			version:       "https://storage.googleapis.com/kubernetes-release-dev/ci/v1.4.0-alpha.2.677+ea69570f61af8e/",
+			expected:      "1.4.0",
+			expectedError: false,
+		},
+		{
+			version:       "",
+			expected:      "",
+			expectedError: true,
+		},
+		{
+			version:       "abc",
+			expected:      "",
+			expectedError: true,
+		},
 	}
-	for version, expected := range grid {
-		sv, err := ParseKubernetesVersion(version)
+	for _, c := range cases {
+		sv, err := ParseKubernetesVersion(c.version)
 		if err != nil {
-			t.Errorf("ParseKubernetesVersion error parsing %q: %v", version, err)
+			if c.expectedError {
+				continue
+			}
+			t.Errorf("ParseKubernetesVersion error parsing %q: %v", c.version, err)
 		}
 
 		actual := sv.String()
-		if actual != expected {
-			t.Errorf("version mismatch: %q -> %q but expected %q", version, actual, expected)
+		if actual != c.expected {
```

Comment:
This will incorrectly pass if the test case expects an error but the function under test returns success and an empty string.

I usually make `expectedError` of type `string` and put the expected error message in it.

Reply:
`Test_TemplateFunctions_CloudControllerConfigArgv` shows a pattern where `expectedError` is an `error`. That appears to be the current pattern in kops.

Reply:
Thanks for your comment, But I can't find where the `Test_TemplateFunctions_CloudControllerConfigArgv` is.

Reply:
upup/pkg/fi/cloudup/template_functions_test.go

Reply:
Thx, done.